### PR TITLE
Docker: add alpine image with Elektra installed

### LINKED ISSUE #3307: Release Alpine image: update Googletest
After the 0.9.2 release, update the official elektra docker image and update to Googletest 1.10.0 (or latest).

Cannot be done before 0.9.2 is released, as 0.9.1 tag does not work with Googletest >= 1.10.0.

### PR DESCRIPTION
**Pull request by mpranj**
Closes #3307.

### CODE REVIEW

File: scripts/docker/alpine/3.10/release.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,75 @@
+FROM alpine:3.10
+
+RUN apk update \
+    && apk add --no-cache --upgrade\
+        bison \
+        build-base \
+        cmake \
+        curl \
+        diffutils \
```
Comment (markus2330) [inline scripts/docker/alpine/3.10/release.Dockerfile]: Why is this needed? libgit2 would be needed for two plugins.

File: scripts/docker/alpine/3.10/release.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        file \
+        git \
+        ninja \
+        ronn \
```
Comment (markus2330) [inline scripts/docker/alpine/3.10/release.Dockerfile]: Is this needed? Does someone wants to compile the man pages in this docker image?

File: scripts/docker/alpine/3.10/release.Dockerfile [hunk 3]
```diff
@@ -0,0 +1,75 @@
+FROM alpine:3.10
+
+RUN apk update \
+    && apk add --no-cache --upgrade\
+        bison \
```
Comment (markus2330) [inline scripts/docker/alpine/3.10/release.Dockerfile]: also flex will be needed for toml

File: scripts/docker/alpine/3.10/release.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        build-base \
+        cmake \
+        curl \
+        diffutils \
+        file \
+        git \
+        ninja \
+        ronn \
+        yaml-cpp-dev
+
```
Comment (markus2330) [inline scripts/docker/alpine/3.10/release.Dockerfile]: Maybe some additional deps like yajl, xerces, augeas would be nice. (If they are available)

File: scripts/docker/alpine/3.10/release.Dockerfile (hunk 1 above)
Comment (mpranj) [reply scripts/docker/alpine/3.10/release.Dockerfile]: I blindly copied the other alpine image and adapted it quickly. :see_no_evil:

File: scripts/docker/alpine/3.10/release.Dockerfile (hunk 2 above)
Comment (mpranj) [reply scripts/docker/alpine/3.10/release.Dockerfile]: Same as above.

File: scripts/docker/alpine/3.10/release.Dockerfile (hunk 1 above)
Comment (mpranj) [reply scripts/docker/alpine/3.10/release.Dockerfile]: Please do note that this is only for building. Then we copy it to a fresh image with only bash installed and the layers are squashed so the image should have minimal size.

Basically my only goal was to provide a minimal image and drop the user into a shell where kdb is available. I do agree that it would be nice to add more plugins for a rich Elektra experience.

Comment (markus2330) [reply scripts/docker/alpine/3.10/release.Dockerfile]: Yes, this can also be done in the next release. I hope everything for the tutorial is included in the current image?

Comment (mpranj) [reply scripts/docker/alpine/3.10/release.Dockerfile]: I did not go through the tutorial. I'll update the image if something is missing. (hopefully alpine has packages for everything needed)

File: scripts/docker/alpine/3.10/release.Dockerfile [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+RUN apk update \
+    && apk add --no-cache --upgrade\
+        bison \
+        build-base \
+        cmake \
+        curl \
+        libgit2 \
+        git \
+        ninja \
+        tcl \
```
Comment (sanssecours) [inline scripts/docker/alpine/3.10/release.Dockerfile]: Are you sure we need [Tcl](https://en.wikipedia.org/wiki/Tcl) for Elektra? As far as I know no part of Elektra requires the programming language. If you want to include the [`tcl` plugin](https://www.libelektra.org/plugins/tcl), then I think adding a package for [Boost](https://www.boost.org) would be the right choice.

Comment (mpranj) [reply scripts/docker/alpine/3.10/release.Dockerfile]: Yes, the idea was to include the tcl plugin as it was part of our asciinema tutorial.

I saw tcl being installed in other docker images so I thought this is needed for the plugin. Thanks for reviewing it, I will add boost.

Further suggestions are absolutely welcome.

**Comment by mpranj** [comment]
I pushed `elektra/elektra:0.9.1-6` to docker hub. Unfortunately boost pulls in python2 and python3 as dependencies, so the image naturally got much larger. (162MB, ~54MB compressed)

It's really a tradeoff, we will have to decide what should be in this image. To me the size is not extreme, but the minimal image had about 7MB.

In return we get a nice list of [plugins](https://github.com/ElektraInitiative/libelektra/files/3908306/plugins.txt).

File: scripts/docker/alpine/3.10/release.Dockerfile (hunk 5 above)
Comment (markus2330) [reply scripts/docker/alpine/3.10/release.Dockerfile]: Yes, this is a common error :wink: Please remove the references of the tcl dependency in the docker images. Elektra never used the tcl interpreter.

**Comment by markus2330** [comment]
Impressive list of plugins :100: 

> I pushed elektra/elektra:0.9.1-6 to docker hub. Unfortunately boost pulls in python2 and python3 as dependencies, so the image naturally got much larger. (162MB, ~54MB compressed)

Unfortunately, the python plugin nevertheless is not included. The tcl plugin is very low priority, so better to not add boost if it pulls python without giving us the python plugin.

> It's really a tradeoff, we will have to decide what should be in this image. To me the size is not extreme, but the minimal image had about 7MB.

I think you balanced it very nicely. Do you know why multifile is not included? https://pkgs.alpinelinux.org/package/edge/main/armv7/fts with 24k sounds to be quite cheap :wink:

**Comment by markus2330** [comment]
Btw. do not waste too much time on fts, maybe we should not use it anyway (and rewrite multifile), see https://wiki.musl-libc.org/faq.html "Why is fts.h not included?"

Let us see if @darddan will find multifile useful for KConfig or if we go with individual mountpoints anyway.

**Comment by mpranj** [comment]
I tried to add `fts` and `fts-dev`. It leads to failures because:
```
Installation: Dlopen failed. Could not load module libelektra-multifile.so. Reason: Error relocating /opt/elektra/build/lib/libelektra-multifile.so: fts_open: symbol not found
```

I don't know how what's going on here, the `LD_LIBRARY_PATH` is set correctly and `nm` says the symbol exists in libfts. Only multifile fails out of all tests.

**Comment by markus2330** [comment]
> I don't know what's going on here, the LD_LIBRARY_PATH is set correctly and nm says the symbol exists in libfts. Only multifile fails out of all tests.

Yes, then simply remove libfts again. Maybe the implementations are somehow incompatible as suggested in the FAQ.

File: scripts/docker/alpine/3.10/release.Dockerfile (hunk 3 above)
Comment (mpranj) [reply scripts/docker/alpine/3.10/release.Dockerfile]: Let's add that when it's needed.

**Comment by markus2330** [comment]
Maybe it makes sense to wait for @Piankero with his demo before merging this.

**Comment by markus2330** [comment]
And it seems like the little patch will not be needed: #3316

**Comment by mpranj** [comment]
> Maybe it makes sense to wait for @Piankero with his demo before merging this.

Sure, let's wait.

> And it seems like the little patch will not be needed

It is needed if we do not publish a new release or change the image to use master instead of the 0.9.1 release. I think for now it's easier to leave it as is until 0.9.2, unless there are some other major fixes.

**Comment by markus2330** [comment]
> It is needed if we do not publish a new release or change the image to use master instead of the 0.9.1 release. I think for now it's easier to leave it as is until 0.9.2, unless there are some other major fixes.

Yes, sounds ok in this case because people trying out the tutorial will most likely use 0.9.2 (as usually only with releases new people start trying out Elektra) or use the docker image.

See also #3319 for a proposal of versioning in future.

**Comment by Piankero** [comment]
I'll write here what actually did not work.

I can user the hosts plugin but I can not add anything to the hosts file:
```sh
sudo kdb mount --with-recommends /etc/hosts system/hosts hosts
sudo kdb set system/hosts/ipv4/my 88.198.134.177
Sorry, 1 warning was issued ;(
	Sorry, module kdb issued the warning C01310:
	Internal: Error during commit. This means backend is broken: system/hosts
Sorry, module resolver_fm_hpu_b issued the error C01100:
Resource: Could not rename file '/etc/hosts.11:1575400385.54897.tmp'. Reason: Resource busy
Create a new key system/hosts/ipv4/my with string "88.198.134.177"
```

this seems to be docker related because I can do it locally

**Comment by markus2330** [comment]
Yes, "Resource busy" sounds like /etc is some overlay-fs or similar. @mpranj any ideas?

**Comment by mpranj** [comment]
Wow, we've managed to hit a very specific error.

In docker you cannot replace bind-mounted files, but you can edit them. Elektra does exactly that, it tries to rename the temp file to the final file name to have a clean commit.

Unfotunately it seems that Docker auto-mounts about 3 files, and `/etc/hosts` is one of them 😢. I'll see if something can be done.

**Comment by markus2330** [comment]
The easiest workaround is to tell Elektra to use a different resolver that does not move files, e.g. wresolver. So `kdb mount -R wresolver ...`.

**Comment by mpranj** [comment]
> The easiest workaround is to tell Elektra to use a different resolver that does not move files, e.g. wresolver. So kdb mount -R wresolver ....

Thank you for this hint. I think it's best to do this, as I did not find any way to tell docker not to bind-mount this file. (putting /etc in a volume also does not help).

**Comment by markus2330** [comment]
@Piankero actually you can use this to show error messages (first mount without sudo, then with wrong resolver, ...) and in between echo to say how you interpret the error message and find the solution.

**Comment by Piankero** [comment]
> actually you can use this to show error messages

this is *such* as specific error which users will not run into (who really mounts the hosts file inside docker which has its own DNS infront?)

> first mount without sudo

I would like to avoid this because the non-sudo error message is very verbose

> then with wrong resolver

The concept of a resolver is already hard to grasp for new users. The fact that there are even multiple resolvers is even more confusing

To me it looks like you urgently want to show the hosts plugin which should be well shown if you mount it, get your localhost IP and try to set a wrong IP. If tutorial users run into this problem here themselves they will also google that this is a non-elektra bug in docker.

**Comment by markus2330** [comment]
> this is such as specific error which users will not run into (who really mounts the hosts file inside docker which has its own DNS infront?).

I agree that it is not a common error.

> Its confusing to be put in a tutorial. Also what if docker ever fixes this issue? Then the tutorial is again outdated

The tutorial needs to be recorded from time to time anyway, also because Elektra changes.

> I would like to avoid this because the non-sudo error message is very verbose

This is strange as there was a PR to specifically fix this error. Why didn't you fix it in a way that you want to show it?

> The concept of a resolver is already hard to grasp for new users. The fact that there are even multiple resolvers is even more confusing

As we see in this case the concept is necessary.

> To me it looks like you urgently want to show the hosts plugin which should be well shown if you mount it, get your localhost IP and try to set a wrong IP.

No, my idea was that the tutorial should also guide through the common/usual errors beginners make. And of course to show many different errors, so that you maybe improve them a bit more.

> If tutorial users run into this problem here themselves they will also google that this is a non-elektra bug in docker.

It is not so easy to understand what is really going on there. Partly because the error messages from the filesystem are really crappy (due to limitations of errno EBUSY is reused in rename for different things).

Anyway, it is okay if you type in something like `kdb mount ... -R wresolver # wresolver doesn't do atomic updates, which is needed because /etc/hosts is a mountpoint in docker`.

This is not only an explanation of why `wresolver` is needed but also that Elektra usually does atomic updates.

**Comment by Piankero** [comment]
> This is strange as there was a PR to specifically fix this error. Why didn't you fix it in a way that you want to show it?

The best case would be to only show that the permission was denied but there are so many warnings gathered via the resolvers which might have other useful messages actually inside them for other non-permission related errors. Thats why I let it as it is and think that users will most likely read the yellow bold part first. For a tutorial though this is way too verbose

> As we see in this case the concept is necessary.

Imo not in an asciinema tutorial but I will simply record what you want ...

> No, my idea was that the tutorial should also guide through the common/usual errors beginners make. And of course to show many different errors, so that you maybe improve them a bit more.

Imo this is done via specification mounting and the 2 errors I make.

> Anyway, it is okay if you type in something like

As you wish

**Comment by Piankero** [comment]
Another problem with the new resolver:

```
elektra $ sudo kdb mount --with-recommends /etc/hosts system/hosts hosts -R wres
olver # wresolver doesn't do atomic updates, which is needed because /etc/hosts 
is a mountpoint in docker
elektra $ sudo kdb get system/hosts/ipv4/localhost
127.0.0.1
 Sorry, 4 warnings were issued ;(
	Sorry, module wresolver issued the warning C01200:
	Installation: Could not retrieve from passwd using getpwuid_r. Could not get ALLUSERSPROFILE for spec, using /
	Sorry, module wresolver issued the warning C01200:
	Installation: Could not get ALLUSERSPROFILE environment variable, using /
	Sorry, module wresolver issued the warning C01200:
	Installation: Could not retrieve from passwd using getpwuid_r. Could not get ALLUSERSPROFILE for spec, using /
	Sorry, module wresolver issued the warning C01200:
	Installation: Could not get ALLUSERSPROFILE environment variable, using /
```

**Comment by markus2330** [comment]
> The best case would be to only show that the permission was denied but there are so many warnings gathered via the resolvers which might have other useful messages actually inside them for other non-permission related errors. Thats why I let it as it is and think that users will most likely read the yellow bold part first. For a tutorial though this is way too verbose

If it works, it also should work in the tutorial. We could also consider to not show warnings if `-v` is not given.

> Imo not in an asciinema tutorial but I will simply record what you want ...

It is only necessary because of the Docker bug. Not showing /etc/hosts validation would be wrong though, as this is one of the killer features of validation.

> Imo this is done via specification mounting and the 2 errors I make.

This is difficult to see from the script. In general a little bit repetition does not harm. Maybe you can somehow divide the whole tutorial in 3 parts to make a better structure.

> Another problem with the new resolver:

Please report a bug, this warning seems to be totally unnecessary as you do not even mount spec. Also for absolute paths this warning makes no sense.

The workaround would be to set the environment variable ALLUSERSPROFILE
@Piankero can you check if this works?

If it works, @mpranj can you do this for the docker container?

**Comment by Piankero** [comment]
> If it works, it also should work in the tutorial. We could also consider to not show warnings if -v is not given.

Yes probably this would be an idea but this would somebody else need to do.

> It is only necessary because of the Docker bug. Not showing /etc/hosts validation would be wrong though, as this is one of the killer features of validation.

as you wish ... but these warnings are not very good in the tutorial ...

> The workaround would be to set the environment variable ALLUSERSPROFILE
@Piankero can you check if this works?

I did `export ALLUSERSPROFILE="elektra"`
before the command and it does not work, still 4 warnings

**Comment by markus2330** [comment]
> Yes probably this would be an idea but this would somebody else need to do.

:cry: 

> as you wish ... but these warnings are not very good in the tutorial ...

Yes, they are not so good in general but if they are there...

> I did export ALLUSERSPROFILE="elektra"
before the command and it does not work, still 4 warnings

Something was done wrongly, maybe the shell does not support `export`? Please do not forget to create the issue, an additional problem is that in line 129 the error message seems to be wrong.

**Comment by mpranj** [comment]
`sudo` intentionally does not preserve the environment. There are options to preserve ENV, but of course it seems like this is not available in sudo of alpine linux.

You can work arount it by:
```
elektra $ sudo ALLUSERSPROFILE="elektra" kdb get system/hosts/ipv4/localhost
Did not find key 'system/hosts/ipv4/localhost'
```
or storing it in `/etc/profile`. It already contains some workarounds that I had to implement, so we can add more.

It seems to me that it still does not help. There is no error but localhost is also not found, even though it is there.

**Comment by Piankero** [comment]
> It seems to me that it still does not help. There is no error but localhost is also not found, even though it is there.

So what is the plan? Should I do the workaround?

**Comment by markus2330** [comment]
> or storing it in /etc/profile. It already contains some workarounds that I had to implement, so we can add more.

Yes, sounds good.

**Comment by mpranj** [comment]
Just to repeat: it seems to me that the workaround does not help. It just leads to different errors such as `Did not find key 'system/hosts/ipv4/localhost'` even though it exists in the config.

**Comment by markus2330** [comment]
`ALLUSERSPROFILE="elektra"` obviously is wrong (I do not know how @Piankero came up with this), for me `ALLUSERSPROFILE=/ kdb get system/hosts/ipv4/localhost` works as expected (without warning).

**Comment by mpranj** [comment]
Ok, that works. I pushed it with `elektra/elektra:0.9.1-8`.

```
elektra $ sudo kdb mount --with-recommends /etc/hosts system/hosts hosts -R wresolver
elektra $ sudo kdb get system/hosts/ipv4/localhost
127.0.0.1
```

**Comment by markus2330** [comment]
Thank you so much!

@Piankero https://www.borgbackup.org/demo.html made quite nice tutorials with more text. (It is maybe already too much text and you need to be quite fast to read all of it.)

File: scripts/docker/alpine/3.10/release.Dockerfile [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        yajl-dev \
+        yaml-cpp-dev \
+        && rm -rf /var/cache/apk/*
+
+RUN echo "%wheel ALL=(ALL) NOPASSWD: ALL" >> /etc/sudoers
+
+RUN echo "alias sudo='sudo -i' # in this image we do not need to be root" >> /etc/profile
+RUN echo "export PS1='\u $ '" >> /etc/profile
+RUN echo "export LD_LIBRARY_PATH=/usr/local/lib/elektra/" >> /etc/profile
+RUN echo "export ALLUSERSPROFILE=/" >> /etc/profile
```
Comment (markus2330) [inline scripts/docker/alpine/3.10/release.Dockerfile]:
```suggestion
RUN echo "export ALLUSERSPROFILE=''" >> /etc/profile
```

an empty string is actually better for the error message:

```
Sorry, 1 warning was issued ;(
        Sorry, module sync issued the warning C01100:
        Resource: Could not open config file //etc/hosts. Reason: Permission denied
Sorry, module hosts issued the error C01100:
Resource: Insufficient permissions to open configuration file //etc/hosts for writing. You might want to retry as root. Reason: Permission denied
```

With my suggestion it is (hopefully, if the syntax is ok) /etc/hosts instead of //etc/hosts.

Comment (mpranj) [reply scripts/docker/alpine/3.10/release.Dockerfile]: Thank you for the improvement! I pushed it to `elektra/elektra:0.9.1-9`.

**Comment by mpranj** [comment]
@markus2330 do you maybe want to build on top of this and update our asciicast? Does not have to be for this release. I can update the docker image if you have further ideas what to add.

You could either make an issue containing a high level "script" of what should be done or directly record the asciicast. I think you have the best feeling for what we should add into our marketing material.

**Comment by markus2330** [comment]
Thank you, yes, we should record an asciicast based on this. We already had a script which was very good, I think we only need to record.

**Comment by mpranj** [comment]
I updated this PR to the current elektra release and pushed it to docker hub (elektra/elektra:0.9.2-1).

I also took the liberty to upgrade our alpine image versions. Let's first see if this does not break our jenkins CI.

**Comment by markus2330** [comment]
Great! Please mention this also in the release notes! :rocket:
